File: java/com/joshmanisdabomb/lcc/registry/LCCEntities.java
package com.joshmanisdabomb.lcc.registry;

import com.joshmanisdabomb.lcc.LCC;
import com.joshmanisdabomb.lcc.entity.*;
import com.joshmanisdabomb.lcc.entity.render.*;
import net.minecraft.entity.EntityClassification;
import net.minecraft.entity.EntityType;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;
import net.minecraftforge.event.RegistryEvent;
import net.minecraftforge.fml.client.registry.RenderingRegistry;

import java.util.ArrayList;

public abstract class LCCEntities {

    public static final ArrayList<EntityType<?>> all = new ArrayList<>();

    //Nostalgic
    public static EntityType<ClassicZombiePigmanEntity> classic_zombie_pigman;

    //Explosives
    public static EntityType<ClassicTNTEntity> classic_tnt;
    public static EntityType<AtomicBombEntity> atomic_bomb;
    public static EntityType<MultipartHitbox> hitbox;
    public static EntityType<NuclearExplosionEntity> nuclear_explosion;

    public static void init(RegistryEvent.Register<EntityType<?>> e) {
        //Helpers
        all.add(hitbox = (EntityType<MultipartHitbox>)EntityType.Builder.<MultipartHitbox>create(MultipartHitbox::new, EntityClassification.MISC).setTrackingRange(10).setUpdateInterval(2).setShouldReceiveVelocityUpdates(true).size(0.98F, 0.98F).setCustomClientFactory((spawn, world) -> new MultipartHitbox(LCCEntities.hitbox, world)).build("hitbox").setRegistryName(LCC.MODID, "hitbox"));

        //Nostalgic
        all.add(classic_zombie_pigman = (EntityType<ClassicZombiePigmanEntity>)EntityType.Builder.<ClassicZombiePigmanEntity>create(ClassicZombiePigmanEntity::new, EntityClassification.MONSTER).setTrackingRange(5).setUpdateInterval(3).setShouldReceiveVelocityUpdates(true).immuneToFire().size(0.6F, 1.95F).build("classic_zombie_pigman").setRegistryName(LCC.MODID, "classic_zombie_pigman"));

        //Explosives
        all.add(classic_tnt = (EntityType<ClassicTNTEntity>)EntityType.Builder.<ClassicTNTEntity>create(ClassicTNTEntity::new, EntityClassification.MISC).setTrackingRange(10).setUpdateInterval(10).setShouldReceiveVelocityUpdates(true).immuneToFire().size(0.98F, 0.98F).setCustomClientFactory((spawn, world) -> new ClassicTNTEntity(LCCEntities.classic_tnt, world)).build("classic_tnt").setRegistryName(LCC.MODID, "classic_tnt"));
        all.add(atomic_bomb = (EntityType<AtomicBombEntity>)EntityType.Builder.<AtomicBombEntity>create(AtomicBombEntity::new, EntityClassification.MISC).setTrackingRange(64).setUpdateInterval(2).setShouldReceiveVelocityUpdates(true).immuneToFire().size(0.98F, 0.98F).setCustomClientFactory((spawn, world) -> new AtomicBombEntity(LCCEntities.atomic_bomb, world)).build("atomic_bomb").setRegistryName(LCC.MODID, "atomic_bomb"));
        all.add(nuclear_explosion = (EntityType<NuclearExplosionEntity>)EntityType.Builder.<NuclearExplosionEntity>create(NuclearExplosionEntity::new, EntityClassification.MISC).setTrackingRange(10).setUpdateInterval(Integer.MAX_VALUE).setShouldReceiveVelocityUpdates(false).size(10F, 10F).setCustomClientFactory((spawn, world) -> new NuclearExplosionEntity(LCCEntities.nuclear_explosion, world)).build("nuclear_explosion").setRegistryName(LCC.MODID, "nuclear_explosion"));
    }

    @OnlyIn(Dist.CLIENT)
    public static void initRenderers() {
        RenderingRegistry.registerEntityRenderingHandler(ClassicTNTEntity.class, manager -> new StateBasedTNTRenderer(LCCBlocks.classic_tnt.getDefaultState(), manager));
        RenderingRegistry.registerEntityRenderingHandler(NuclearExplosionEntity.class, NuclearExplosionRenderer::new);
        RenderingRegistry.registerEntityRenderingHandler(ClassicZombiePigmanEntity.class, ClassicZombiePigmanRenderer::new);
        RenderingRegistry.registerEntityRenderingHandler(MultipartHitbox.class, EmptyRenderer::new);
        RenderingRegistry.registerEntityRenderingHandler(AtomicBombEntity.class, AtomicBombRenderer::new);
    }

}

File: lcc-content/src/main/kotlin/com/joshmanisdabomb/lcc/directory/LCCEntities.kt
package com.joshmanisdabomb.lcc.directory

import com.joshmanisdabomb.lcc.LCC
import com.joshmanisdabomb.lcc.entity.*
import com.joshmanisdabomb.lcc.entity.render.*
import com.joshmanisdabomb.lcc.facade.boat.LCCBoatEntity
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.fabricmc.fabric.api.`object`.builder.v1.entity.FabricDefaultAttributeRegistry
import net.fabricmc.fabric.api.`object`.builder.v1.entity.FabricEntityTypeBuilder
import net.fabricmc.fabric.api.client.rendering.v1.EntityRendererRegistry
import net.minecraft.client.render.entity.BoatEntityRenderer
import net.minecraft.client.render.entity.EmptyEntityRenderer
import net.minecraft.client.render.entity.SkeletonEntityRenderer
import net.minecraft.entity.*
import net.minecraft.entity.mob.HostileEntity
import net.minecraft.entity.mob.MobEntity
import net.minecraft.util.registry.Registry
import net.minecraft.world.Heightmap

object LCCEntities : AdvancedDirectory<FabricEntityTypeBuilder<out Entity>, EntityType<out Entity>, Unit, Unit>(), RegistryDirectory<EntityType<out Entity>, Unit, Unit> {

    override val registry by lazy { Registry.ENTITY_TYPE }

    override fun regId(path: String) = LCC.id(path)

    val pocket_zombie_pigman by entry(::typeInitialiser) { FabricEntityTypeBuilder.createMob<PocketZombiePigmanEntity>().spawnGroup(SpawnGroup.MONSTER).entityFactory(::PocketZombiePigmanEntity).dimensions(EntityDimensions.fixed(0.6f, 1.95f)).fireImmune().trackRangeChunks(5).trackedUpdateRate(3).forceTrackedVelocityUpdates(true) }
        .addInitListener { context, params -> FabricDefaultAttributeRegistry.register(context.entry, PocketZombiePigmanEntity.createAttributes()) }

    val classic_tnt by entry(::typeInitialiser) { FabricEntityTypeBuilder.create(SpawnGroup.MISC, ::ClassicTNTEntity).dimensions(EntityDimensions.fixed(0.98f, 0.98f)).fireImmune().trackRangeChunks(10).trackedUpdateRate(10).forceTrackedVelocityUpdates(true) }

    val atomic_bomb by entry(::typeInitialiser) { FabricEntityTypeBuilder.create(SpawnGroup.MISC, ::AtomicBombEntity).dimensions(EntityDimensions.fixed(0.98f, 0.98f)).fireImmune().trackRangeChunks(100).trackedUpdateRate(2).forceTrackedVelocityUpdates(true) }
    val nuclear_explosion by entry(::typeInitialiser) { FabricEntityTypeBuilder.create(SpawnGroup.MISC, ::NuclearExplosionEntity).dimensions(EntityDimensions.fixed(0.1f, 0.1f)).fireImmune().trackRangeChunks(100).trackedUpdateRate(1).forceTrackedVelocityUpdates(false) }

    val salt by entry(::typeInitialiser) { FabricEntityTypeBuilder.create(SpawnGroup.MISC, ::SaltEntity).dimensions(EntityDimensions.fixed(0.25f, 0.25f)).trackRangeChunks(4).trackedUpdateRate(10).forceTrackedVelocityUpdates(true) }
    val consumer_tongue by entry(::typeInitialiser) { FabricEntityTypeBuilder.create(SpawnGroup.MISC, ::ConsumerTongueEntity).dimensions(EntityDimensions.fixed(0.5f, 0.5f)).trackRangeChunks(4).trackedUpdateRate(10).forceTrackedVelocityUpdates(true) }
    val disciple_dust by entry(::typeInitialiser) { FabricEntityTypeBuilder.create(SpawnGroup.MISC, ::DiscipleDustEntity).dimensions(EntityDimensions.fixed(0.6f, 0.6f)).trackRangeChunks(4).trackedUpdateRate(10).forceTrackedVelocityUpdates(true) }

    val traveller by entry(::typeInitialiser) { FabricEntityTypeBuilder.createMob<TravellerEntity>().spawnGroup(SpawnGroup.CREATURE).entityFactory(::TravellerEntity).dimensions(EntityDimensions.changing(0.6f, 1.95f)).trackRangeChunks(10).trackedUpdateRate(3).forceTrackedVelocityUpdates(true) }
        .addInitListener { context, params -> FabricDefaultAttributeRegistry.register(context.entry, MobEntity.createMobAttributes()) }

    val consumer by entry(::typeInitialiser) { FabricEntityTypeBuilder.createMob<ConsumerEntity>().spawnGroup(SpawnGroup.MONSTER).entityFactory(::ConsumerEntity).dimensions(EntityDimensions.changing(0.8f, 1.1f)).trackRangeChunks(8).trackedUpdateRate(3).forceTrackedVelocityUpdates(true).spawnRestriction(SpawnRestriction.Location.ON_GROUND, Heightmap.Type.MOTION_BLOCKING_NO_LEAVES, LCCSpawnRestrictions::canSpawnInDarkOrSkylight) }
        .addInitListener { context, params -> FabricDefaultAttributeRegistry.register(context.entry, ConsumerEntity.createAttributes()) }
    val wasp by entry(::typeInitialiser) { FabricEntityTypeBuilder.createMob<WaspEntity>().spawnGroup(SpawnGroup.MONSTER).entityFactory(::WaspEntity).dimensions(EntityDimensions.changing(0.99f, 0.675f)).trackRangeChunks(5).trackedUpdateRate(3).forceTrackedVelocityUpdates(true).spawnRestriction(SpawnRestriction.Location.ON_GROUND, Heightmap.Type.MOTION_BLOCKING_NO_LEAVES, LCCSpawnRestrictions::canSpawnInSkylight) }
        .addInitListener { context, params -> FabricDefaultAttributeRegistry.register(context.entry, WaspEntity.createAttributes()) }
    val baby_skeleton by entry(::typeInitialiser) { FabricEntityTypeBuilder.createMob<BabySkeletonEntity>().spawnGroup(SpawnGroup.MONSTER).entityFactory(::BabySkeletonEntity).dimensions(EntityDimensions.changing(0.6f, 1.99f)).trackRangeChunks(8).trackedUpdateRate(3).forceTrackedVelocityUpdates(true).spawnRestriction(SpawnRestriction.Location.ON_GROUND, Heightmap.Type.MOTION_BLOCKING_NO_LEAVES, HostileEntity::canSpawnInDark) }
        .addInitListener { context, params -> FabricDefaultAttributeRegistry.register(context.entry, BabySkeletonEntity.createAttributes()) }
    val disciple by entry(::typeInitialiser) { FabricEntityTypeBuilder.createMob<DiscipleEntity>().spawnGroup(SpawnGroup.MONSTER).entityFactory(::DiscipleEntity).dimensions(EntityDimensions.changing(0.8f, 2.15f)).trackRangeChunks(8).trackedUpdateRate(3).forceTrackedVelocityUpdates(true).spawnRestriction(SpawnRestriction.Location.ON_GROUND, Heightmap.Type.MOTION_BLOCKING_NO_LEAVES, LCCSpawnRestrictions::canSpawnInDarkOrSkylight) }
        .addInitListener { context, params -> FabricDefaultAttributeRegistry.register(context.entry, DiscipleEntity.createAttributes()) }
    val psycho_pig by entry(::typeInitialiser) { FabricEntityTypeBuilder.createMob<PsychoPigEntity>().spawnGroup(SpawnGroup.MONSTER).entityFactory(::PsychoPigEntity).dimensions(EntityDimensions.changing(0.9f, 0.9f)).trackRangeChunks(10).trackedUpdateRate(3).forceTrackedVelocityUpdates(true).spawnRestriction(SpawnRestriction.Location.ON_GROUND, Heightmap.Type.MOTION_BLOCKING_NO_LEAVES, LCCSpawnRestrictions::canSpawnInDarkOrSkylight) }
        .addInitListener { context, params -> FabricDefaultAttributeRegistry.register(context.entry, PsychoPigEntity.createAttributes()) }
    val rotwitch by entry(::typeInitialiser) { FabricEntityTypeBuilder.createMob<RotwitchEntity>().spawnGroup(SpawnGroup.MONSTER).entityFactory(::RotwitchEntity).dimensions(EntityDimensions.changing(1.2f, 0.8f)).trackRangeChunks(8).trackedUpdateRate(3).forceTrackedVelocityUpdates(true).spawnRestriction(SpawnRestriction.Location.ON_GROUND, Heightmap.Type.MOTION_BLOCKING_NO_LEAVES, LCCSpawnRestrictions::canSpawnInDarkOrSkylight) }
        .addInitListener { context, params -> FabricDefaultAttributeRegistry.register(context.entry, RotwitchEntity.createAttributes()) }
    val fly by entry(::typeInitialiser) { FabricEntityTypeBuilder.createMob<FlyEntity>().spawnGroup(SpawnGroup.MONSTER).entityFactory(::FlyEntity).dimensions(EntityDimensions.changing(0.3f, 0.3f)).trackRangeChunks(5).trackedUpdateRate(3).forceTrackedVelocityUpdates(true) }
        .addInitListener { context, params -> FabricDefaultAttributeRegistry.register(context.entry, FlyEntity.createAttributes()) }
    val woodlouse by entry(::typeInitialiser) { FabricEntityTypeBuilder.createMob<WoodlouseEntity>().spawnGroup(SpawnGroup.CREATURE).entityFactory(::WoodlouseEntity).dimensions(EntityDimensions.changing(0.8f, 0.625f)).trackRangeChunks(10).trackedUpdateRate(3).forceTrackedVelocityUpdates(true).spawnRestriction(SpawnRestriction.Location.ON_GROUND, Heightmap.Type.MOTION_BLOCKING_NO_LEAVES, LCCSpawnRestrictions::canSpawnInSkylight) }
        .addInitListener { context, params -> FabricDefaultAttributeRegistry.register(context.entry, WoodlouseEntity.createAttributes()) }

    val rubber_boat: EntityType<LCCBoatEntity> get() = LCCBoatTypes.rubber.entityType
    val deadwood_boat: EntityType<LCCBoatEntity> get() = LCCBoatTypes.deadwood.entityType

    private fun <E : Entity> typeInitialiser(input: FabricEntityTypeBuilder<E>, context: DirectoryContext<Unit>, parameters: Unit): EntityType<E> {
        return initialiser(input.build(), context, parameters)
    }

    override fun defaultProperties(name: String) = Unit
    override fun defaultContext() = Unit

    @Environment(EnvType.CLIENT)
    fun initRenderers() {
        EntityRendererRegistry.register(pocket_zombie_pigman, ::PocketZombiePigmanEntityRenderer)

        EntityRendererRegistry.register(traveller, ::TravellerEntityRenderer)

        EntityRendererRegistry.register(consumer, ::ConsumerEntityRenderer)
        EntityRendererRegistry.register(wasp, ::WaspEntityRenderer)
        EntityRendererRegistry.register(baby_skeleton, ::SkeletonEntityRenderer)
        EntityRendererRegistry.register(disciple, ::DiscipleEntityRenderer)
        EntityRendererRegistry.register(psycho_pig, ::PsychoPigEntityRenderer)
        EntityRendererRegistry.register(rotwitch, ::RotwitchEntityRenderer)
        EntityRendererRegistry.register(fly, ::FlyEntityRenderer)
        EntityRendererRegistry.register(woodlouse, ::WoodlouseEntityRenderer)

        EntityRendererRegistry.register(atomic_bomb, ::AtomicBombEntityRenderer)
        EntityRendererRegistry.register(nuclear_explosion, ::NuclearExplosionEntityRenderer)
        EntityRendererRegistry.register(salt, ::SaltEntityRenderer)
        EntityRendererRegistry.register(consumer_tongue, ::ConsumerTongueEntityRenderer)
        EntityRendererRegistry.register(disciple_dust, ::EmptyEntityRenderer)

        EntityRendererRegistry.register(classic_tnt) { dispatcher -> StateBasedTNTEntityRenderer(LCCBlocks.classic_tnt.defaultState, dispatcher) }

        EntityRendererRegistry.register(rubber_boat) { dispatcher -> BoatEntityRenderer(dispatcher, false) }
        EntityRendererRegistry.register(deadwood_boat) { dispatcher -> BoatEntityRenderer(dispatcher, false) }
    }

}